Generalised oscillator strengths in GOSH format

gguzzina opened the pull request:
As discussed in #2815 a new dataset of Generalised Oscillator Strengths is now available at https://zenodo.org/record/6599071
Using this file finally removes the reliance on the older dataset formerly available through DM

Changes:
- add support for GOSH files (documented at https://gitlab.com/gguzzina/gosh) allowing to use the new dataset, and 
- The new GOS are implemented in a new file `gosh_gos.py`
- They can be selected by hand by using `create_model(GOS='gosh')`
- The new format is default, when available. `EELSSpectrum.create_model()` will try, in order, to use GOSH, then Hartree-Slater, then Hydrogenic.
- The path for the database file is set in `preferences.EELS.eels_gosh_database_path`

I based this on RELEASE_next_minor because the changes are very contained, and should not break backwards compatibility.

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring,
- [x] update user guide,
- [x] add an changelog entry in the `upcoming_changes` folder,
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np
hl = hs.datasets.eelsdb(title="Hexagonal Boron Nitride", spectrum_type="coreloss")[0]
hl.add_elements(('B', 'N'))
hl.crop_signal1D(150.0, 610.0)

m = hl.create_model(GOS='gosh')
m.fit()
m.plot()
```

gguzzina committed:
Add initial support for GOS5 files.

Add support for Generalised oscillator strengths databases in the
GOS5 file format, documented at https://gitlab.com/gguzzina/gos5

These can be used by using GOS='GOS' as an argument to create_model()

GOS5 databases are the new default, when available.
EELSSpectrum.create_model() will try, in order, to use
 GOS5, then Hartree-Slater, then Hydrogenic.

gguzzina commented:
@ericpre no pooch for now. That will have to wait.

ericpre commented on hyperspy/misc/eels/hdf5_gos.py:
```diff
@@ -0,0 +1,188 @@
+
+class Hdf5GOS(GOSBase):
+
+    """Read Hartree-Slater Generalized Oscillator Strenght parametrized
```
Typo?

gguzzina replied:
Well, of course I started by copying the HS file. I clearly did not look carefully enough for instances of the name.

ericpre commented on hyperspy/misc/eels/hdf5_gos.py:
```diff
@@ -0,0 +1,188 @@
+R = constants.value("Rydberg constant times hc in eV")
+a0 = constants.value("Bohr radius")
+
+class Hdf5GOS(GOSBase):
```
```suggestion
class DFTGOS(GOSBase):
```
?

ericpre commented on hyperspy/misc/eels/hdf5_gos.py:
```diff
@@ -0,0 +1,188 @@
+
+    """
+
+    _name = 'Hartree-Slater'
```
It needs updating.

ericpre commented on hyperspy/_components/eels_cl_edge.py:
```diff
@@ -55,8 +63,9 @@ class EELSCLEdge(Component):
         GOS was exported using `GOS.as_dictionary`, and will be reconstructed.
-    GOS : {'hydrogenic', 'Hartree-Slater', None}
-        The GOS to use. If None it will use the Hartree-Slater GOS if
+    GOS : {'hydrogenic', 'GOS5', 'Hartree-Slater', None}
```
```suggestion
    GOS : {'hydrogenic', 'DFT', 'Hartree-Slater', None}
```

ericpre reviewed (commented):
This looks good, it simply need some test in https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/tests/model/test_eelsmodel.py#L61!
To download the file, you can use something along the line of https://github.com/hyperspy/hyperspy/blob/f3e28eb4694ccdc46e38a841348d3658b920cb1f/hyperspy/tests/io/test_edax.py#L15-L40:

I suggested to use `DFT` instead of `GOS5` but there may be better name.

> @ericpre no pooch for now. That will have to wait.

No worries, we can do this in a follow up PR!

gguzzina commented:
> I suggested to use `DFT` instead of `GOS5` but there may be better name.

I had actually started with that, but in the end I'm implementing compatibility for a file format, with no assumption on how the underlying data has been computed. While ours obviously have been computed by DFT, others may differ.

I don't particularly like GOS5 though. Suggestions are welcome.

gguzzina committed:
Fix Saving to file of GOS5 edges, update docstrings

Saving a model to file now correctly exports GOS5 edges,
including the DOI of the used GOS5 dataset.
Loading also appears to work.

Fixed several typos and PEP violations.

francisco-dlp commented:
Glad to see this moving forward. Regarding the name, what about GOSH?

gguzzina commented:
GOSH is clever!
I think I could go with that, at the cost of changing it also in the Zenodo description, and the specifications repo.
Well, I'm computing a new version of the database anyway to include high loss edges.

ericpre commented:
> > I suggested to use `DFT` instead of `GOS5` but there may be better name.
> 
> I'd actually started with that, but in the end I'm implementing compatibility for a file format, with no assumption on how the underlying data has been computed. While ours obviously have been computed by DFT, others may differ.

Indeed in this case, there is two approaches you can take:
1. Have a base class (`Hdf5GOS`) supporting the format for the GOS and a child class `GOSH` or `GOS_DFT`, which is specific to this type of GOS and will know the path to the correct file. If one day, someone add another type of GOS, then another child class of `Hdf5GOS` can be created. In the [original issue](https://github.com/hyperspy/hyperspy/issues/2815#issuecomment-1165740767), @zezhong-zhang mentioned that they would interested in using it too, so that it may well happen sooner than later! ;)
2. Don't plan for hypothetical future addition and keep it as it is but with a name describing what it is instead of its format to be consistent with the rest of the code, e.g. `HartreeSlaterGOS` and `HydrogenicGOS`

> I don't particularly like GOS5 though. Suggestions are welcome.

To be fair, you are creating these GOS, so it is your call! The reason I suggested `DFT` is to be consistent with `Hydrogenic` and `Hartree-Slater`.

zezhong-zhang commented:
> 1. Have a base class (`Hdf5GOS`) supporting the format for the GOS and a child class `GOSH` or `GOS_DFT`, which is specific to this type of GOS and will know the path to the correct file. If one day, someone add another type of GOS, then another child class of `Hdf5GOS` can be created. In the [original issue](https://github.com/hyperspy/hyperspy/issues/2815#issuecomment-1165740767), @zezhong-zhang mentioned that they would interested in using it too, so that it may well happen sooner than later! ;)

@ericpre Yes, we can take the `DiracGOS` child class or `GOS_Dirac` to be consistent. The calling should be something like 

```python
m = hl.create_model(GOS='Dirac')
```

Our database will be released maybe in Feb. Will update the link in the community & pull request to get it supported. 

Good work  @gguzzina,  glad to see the integration with hyperspy!

gguzzina commented:
I'll go with some form of option 1 then, it seems the cleanest option for the moment.

I have a few ideas on how to make it all tidy again, in the unlikely event that GOS datasets were to suddenly proliferate.

gguzzina commented:
@ericpre @francisco-dlp 

On second thought, I don't think there should not be a class for every GOS available. I think there should be one class, supporting the file format, and the user should have an easy way to use any dataset they want. After all the entire point of making a standard file format is to have interoperability without changing the underlying software.

This should include the possibility to supply a file path, or to have one of these datasets downloaded for them using e.g. pooch.
Then all public ones can be included in a list of downloadable files that ships with hyperspy, and adding a new one is as hard as giving a file path.

Thoughts?

edit: I will however be using GOSH as the name for the file format.

ericpre commented:
Yes, this sounds good!

gguzzina committed:
File format renamed to GOSH

gguzzina committed:
Add test for gosh format GOS

gguzzina committed:
Fix load/save test failure

Save/Load model was failing because the subshell_factor was not
stored for GOSH files

Skip the GOSH testing if files can't be downloaded.

gguzzina committed:
Improve GOSH related docstring and preferences

Remove references to DFT from docstrings

Set default for preferences.EELS.eels_gosh_database_path and improve
 the label

gguzzina committed:
Add changelog and documentation for PR #3082

gguzzina committed:
Make documentation for #3082 more readable

Proper formatting, and clearer phrasing.

ericpre reviewed (approved):
I didn't check the math in the `integrateq` method but the rest looks good to me!

gguzzina commented:
> I didn't check the math in the `integrateq` method but the rest looks good to me!

Well, that's a 100% copy-paste from hartree_slater_gos.py

gguzzina committed:
Fix: The subshell factor was applied twice

gguzzina commented:
The last commit corrects an error on my side, in the loading code. A normalisation factor was applied twice, but this would only become visible when the factor was different from 1.

I also published a new version of the dataset, including more edges. I will update the tests accordingly.

ericpre commented:
> > I didn't check the math in the `integrateq` method but the rest looks good to me!
> 
> Well, that's a 100% copy-paste from hartree_slater_gos.py

Ok, since the only method which is different is `readgosarray`/`readgosfile`, would it make sense to make another base class which contains all common methods and the child classes would only contain the method to read the GOS data?

gguzzina commented:
> Ok, since the only method which is different is `readgosarray`/`readgosfile`, would it make sense to make another base class which contains all common methods and the child classes would only contain the method to read the GOS data?

If we are doing that, we might as well make a few minor changes to HydrogenicGOS and merge that one in as well.

gguzzina committed:
Merge branch 'RELEASE_next_minor' into gos5_gos

gguzzina committed:
Update GOS dataset version

gguzzina committed:
Consolidate intergrateq function into base_gos

gosh_gos.py and hartree_slater_gos.py contained identical methods integrateq.
These have been moved to the base class base_gos.py

gguzzina commented:
I moved the integrateq method, present in both hartree_slater_gos and gosh_gos, into base_gos, the base class for all GOS.

hydrogenic_gos is just different enough that it's not trivial to include it without significant restructuring, and after two weeks of not doing it I just have to admit that I just don't have the time right now.

All tests still pass, but I'll conduct a few more tests by hand to ensure everything still works as before.

ericpre committed:
Add `TabulatedGOS` class to share code between `HartreeSlaterGOS` and `GoshGOS`

ericpre committed:
Tweak documentation

ericpre committed:
Use pooch to download GOSH dataset from zenodo

ericpre committed:
Add `gos_file_path` to specify file path of gosh file to use in EELSModel

ericpre committed:
Add more `binding_energy` in the `elements` database and update tests

ericpre committed:
Rename `GOSBase` to `BaseGOS`

ericpre committed:
Increase coverage of EELS GOS

ericpre committed:
Improve EELS documentation and add example showing how to use custom GOS

ericpre commented:
I tidied up a few things in this PR and got carried away:
- using pooch to download the data
- improving the structure of the class
- improving the coverage
- extending the `binding_energy` in the `elements` database to support the additional edges available in https://zenodo.org/record/7645765
- adding support for using different GOS from gosh file, see https://hyperspy--3082.org.readthedocs.build/en/3082/user_guide/eels.html#generalised-oscillator-strengths

@gguzzina, do you want to review my changes?

gguzzina commented:
@ericpre I'm happy to review your changes. Is there any urgency to this?
Right now I'm in over my head, but I also don't want to keep you guys waiting.

ericpre commented:
No there is no urgency. Do you have an approximate idea of when you would be able to review it?

gguzzina commented:
I'll try to find a little bit of time this week, otherwise week 23.

gguzzina commented:
I've gone through the changes, and it looks good for me!

ericpre commented:
Thanks @gguzzina, it is very nice to see this implemented!
